### PR TITLE
Faucet can only be used once a day per address

### LINKED ISSUE #962: Faucet limit from draining
Hey,

Is there currently a way in the faucet to limit faucet users from draining it, like IP rate limiter per day or something like that, which would prevent somebody from spamming the faucet?

### PR DESCRIPTION
**Pull request by msteiner96**
Resolves #962

### CODE REVIEW

File: packages/faucet/src/api/webserver.ts [hunk 1]
```diff
@@ -14,8 +14,14 @@ export interface ChainConstants {
   readonly chainId: string;
 }
 
+export interface AddressEntry {
+  address: string;
+  date: number;
```
Comment (webmaster128) [inline packages/faucet/src/api/webserver.ts:19]: I think this makes the code a bit more intuitive to read

```suggestion
  lastUsed: Date;
```

File: packages/faucet/src/api/webserver.ts [hunk 2]
```diff
@@ -73,6 +88,8 @@ export class Webserver {
 
           try {
             await faucet.credit(address, matchingDenom);
+            // Count addresses to prevent draining
+            this.addressCounter.push({ address: address, date: Date.now() });
```
Comment (webmaster128) [inline packages/faucet/src/api/webserver.ts:92]: This adds the same address multiple times to the array. But `const addressUsed = this.addressCounter.find((x) => x.address === address);` will only search for the first entry. So after 24h, no request will be blocked anymore.

We probably make our lives easier using a Map instead.

File: packages/faucet/src/api/webserver.ts [hunk 3]
```diff
@@ -58,21 +59,32 @@ export class Webserver {
...
+          if (entry !== undefined) {
+            if (entry.getDate() + 24 * 3600 > Date.now()) {
+              throw new HttpError(
+                405,
+                "Too many request from the same address. Blocked to prevent draining. Please wait 24h and try it again!",
+              );
+            }
+          }
+
+          const availableTokens: string[] = await faucet.availableTokens();
```
Comment (webmaster128) [inline packages/faucet/src/api/webserver.ts:78]: No need to change this line as part of this PR, right?

```suggestion
          const availableTokens = await faucet.availableTokens();
```

File: packages/faucet/src/api/webserver.ts [hunk 4]
```diff
@@ -58,21 +59,32 @@ export class Webserver {
...
           const creditBody = RequestParser.parseCreditBody(requestBody);
-
           const { address, denom } = creditBody;
 
           if (!isValidAddress(address, constants.addressPrefix)) {
             throw new HttpError(400, "Address is not in the expected format for this chain.");
           }
 
-          const availableTokens = await faucet.availableTokens();
+          const entry: Date | undefined = this.addressCounter.get(address);
```
Comment (webmaster128) [inline packages/faucet/src/api/webserver.ts:68]: Do you need to make this type explicit? It should be infered to `Date | undefined`

File: packages/faucet/src/api/webserver.ts (hunk 3 above)
Comment (msteiner96) [reply packages/faucet/src/api/webserver.ts:78]: Was my IDE somehow... Gonna delete it

File: packages/faucet/src/api/webserver.ts [hunk 5]
```diff
@@ -58,13 +59,22 @@ export class Webserver {
...
-
           const { address, denom } = creditBody;
 
           if (!isValidAddress(address, constants.addressPrefix)) {
             throw new HttpError(400, "Address is not in the expected format for this chain.");
           }
 
+          const entry = this.addressCounter.get(address);
+          if (entry !== undefined) {
+            if (entry.getDate() + 24 * 3600 > Date.now()) {
```
Comment (webmaster128) [inline packages/faucet/src/api/webserver.ts:70]: >  The getDate() method returns the day of the month for the specified date according to local time. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getDate

Seems like this is not what we want here

**Review by webmaster128 — COMMENTED** [review]
Could you add a CHANGELOG entry?

File: packages/faucet/src/api/webserver.ts (hunk 5 above)
Comment (msteiner96) [reply packages/faucet/src/api/webserver.ts]: I'm stupid. Got confused. It's `getTime()` ofc

**Comment by msteiner96** [comment]
> Could you add a CHANGELOG entry?

-- Done

**Review by webmaster128 — APPROVED** [review]

**Pull request merged**